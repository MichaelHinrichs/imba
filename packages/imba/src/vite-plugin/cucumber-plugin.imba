import compiler from "gherking";
import { toRollupError } from "./utils/error";
import {
	preResolveOptions,
	resolveOptions,
} from "./utils/options";
import { createCompileImba } from "./utils/compile";
import Handlebars from "handlebars";
import rawTestTemplate from './cucumberTemplate.txt?raw'

Handlebars.registerHelper('isRuleOrScenario') do(a) a == 'Scenario' or a == 'Rule'
Handlebars.registerHelper('isScenarioOutline') do(a) a == 'Scenario Outline' or a == 'Scenario Template'
Handlebars.registerHelper('getDescriptionFromCells') do(cells)
	cells.map(do $1.value).join(', ')
Handlebars.registerHelper('getCellValue') do(header, index)
	header.cells[index].value
###
{{> scenario item=. parentId=../this.id }}
###
def indent(str, times = 1)
	let indentation = ""
	indentation += "\t" for i in [0 ... times]
	str
		.split('\n')
		.map(do "{indentation}{$1}")
		.join('\n')
const scenarioText = '''
s = global.Steps.find('{{{step.text}}}', '{{ step.keyword }}').stepDefinition
await s.target[s.fname]
	.apply(
		contexts.c{{ parentIndex }}.context,
		# variables
		[s.cucumberExpression.match('{{{step.text}}}').map(do $1.getValue!)]
	)
'''

Handlebars.registerPartial('step', "\n{indent(scenarioText, 2)}\n")
const L = console.log

const testTemplate = Handlebars.compile(rawTestTemplate)

export def generateImbaCode(id)
	let ast = await compiler.load(id)
	# ast = await compiler.process(ast)
	const doc = ast[0]
	const feature = doc.feature
	# L feature
	# debugger
	const stepDefsGlob = './step_definitions/**/*.imba'
	const backgroundEl = feature.elements.filter(do $1.keyword == 'Background')[0]
	const baseContextPath = './context.imba'

	def hasDecoractor(item, val)
		return yes if item.tags..find(do $1.name == val)
		for el in item.elements or []
			return hasDecoractor(el)
		no
	const code = testTemplate({
		feature
		backgroundEl
		stepDefsGlob
		baseContextPath
		hasNoTimeout: hasDecoractor(feature, 'notimeout')
	})

def _transform(id, compileImba, options)
	const code = await generateImbaCode(id)
	let compiledData
	const imbaRequest = {
		id: id
		filename: id
		normalizedFilename: id
		query: {}
	}
	try compiledData = await compileImba(imbaRequest, code, options) catch e
		throw toRollupError(e, options);
	return compiledData.compiled.js

export default def cucumberPlugin(inlineOptions = {})
	let options
	let compileImba\( (imbaRequest: ImbaRequest,code: string,options: Partial<ResolvedOptions>,) => Promise<CompileData>)

	
	def config(config, configEnv)
		options = await preResolveOptions(inlineOptions, config, configEnv)

	def configResolved(config)
		options = resolveOptions(options, config);
		# patchResolvedViteConfig(config, options);
		compileImba = createCompileImba(options);
		
	def transform(_, id)
		if id.endsWith('.feature')
			await _transform(id, compileImba, options)

	return 
		name: "vite-plugin-imba-cucumber"
		enforce: 'pre'
		config: config
		configResolved: configResolved
		transform: transform

				
