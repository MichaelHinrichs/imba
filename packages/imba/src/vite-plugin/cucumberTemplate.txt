const modules = import.meta.glob('{{stepDefsGlob}}', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
const decorators = {
}

{{#if hasNoTimeout}}
vi.setConfig(testTimeout: 50000000)
{{/if}}

beforeAll do(suite)
	{{#each feature.tags}}
	decorators['{{this.name}}'] = '{{this.value}}' or yes
	   
	{{/each}}
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	{{#each feature.elements}}
	elName = '{{{ this.keyword }}}: {{{ this.name }}}'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		{{#each this.tags}}
		suite.meta.decorators['{{this.name}}'] = '{{this.value}}' or yes
		
		{{/each}}
		{{!-- L "creating context c{{ @index }}" --}}
		contexts.c{{ @index }} = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	{{/each}}

	{{#if backgroundEl}}
	   {{#each backgroundEl.steps}}
	s = global.Steps.find('{{{this.text}}}', '{{ this.keyword}}').stepDefinition
	# find the context for the current test this beforeEach hook is running before
	c = Object.keys(contexts).find(do 
		contexts[$1].name == suite.meta.name or contexts[$1].name == suite.meta.suite..suite..name 
	)
	# apply the step
	await s.target[s.fname].apply(
		contexts[c].context,
		[s.cucumberExpression.match('{{{this.text}}}').map(do $1.getValue!)]
	)
	{{!-- L "applyed background on {c}", contexts[c].context.result --}}

	   {{/each}}
	{{/if}}

describe 'Feature: {{ feature.name }}' do
{{#each feature.elements}}
	{{#if (isScenarioOutline this.keyword)}}
	describe '{{ this.keyword }}: {{ this.name }}' do
		{{#each this.examples}}
		describe 'Examples group {{ @index }}: {{ this.name}}' do
			let text
			{{#each this.body}}
			test 'Example {{ @index }}: {{getDescriptionFromCells this.cells}}' do
				{{#each ../../this.steps}}
				text = '{{{ this.text }}}'
				{{#each ../this.cells}}
				text = text.replace('<{{getCellValue ../../../this.header @index}}>', {{this.value}})
				{{/each}}
				s = global.Steps.find(text, '{{ this.keyword }}').stepDefinition
				{{!-- L "running step definition for example {{ @../index }}: {text}"
				L contexts
				L '{{ @../index}}, {{ @../../index}}, {{ @../../../index}}, {{ @../../../../index}}'
				L contexts.c{{ @../../../index}}.result --}}
				await s.target[s.fname]
					.apply(
						contexts.c{{ @../../../index}}.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				{{/each}}
			{{/each}}
		{{/each}}
	{{/if}}
	{{#if (isRuleOrScenario this.keyword)}}
	test('{{ this.keyword }}: {{ this.name }}')	do
		{{#each this.steps}}
{{> step step=. parentIndex=@../index }}
		{{/each}}
	{{/if}}
{{/each}}