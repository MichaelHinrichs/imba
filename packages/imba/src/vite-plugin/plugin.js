
/*body*/

import svgPlugin from "./svg-plugin";;
;
import {transformWithEsbuild} from 'vite';
import {buildIdParser,normalize,injectQuery} from "./utils/id";;
;
import {log,logCompilerWarnings} from "./utils/log";;
import {createCompileImba} from "./utils/compile";;
;
import {buildExtraViteConfig,patchResolvedViteConfig,preResolveOptions,resolveOptions,validateInlineOptions} from "./utils/options";;

import {toRollupError} from "./utils/error";;
;
import {resolveViaPackageJsonImba} from "./utils/resolve";;
import {ensureWatchedFile,setupWatchers} from "./utils/watch";;
import {handleImbaHotUpdate} from './handle-imba-hot-update';;
import url from 'url';

const allCssModuleId = 'virtual:imba/*?css';
const resolvedAllCssModuleId = ("\0" + allCssModuleId);

/**
@param {Partial<Options>} inlineOptions
*/
export default function imbaPlugin(inlineOptions = {}){
	
	let name = "vite-plugin-imba";
	let enforce = "pre";
	let requestParser;
	let compileImba;
	let options;
	let viteConfig;
	let api = {};
	let resolvedImbaSSR;
	let testΦ;
	let buildΦ;
	let devΦ;
	
	validateInlineOptions(inlineOptions);
	const cache = new VitePluginImbaCache();;
	
	const plugins = [];
	async function config(config,configEnv){
		
		if (process.env.DEBUG) {
			
			log.setLevel("debug");
		} else if (config.logLevel) {
			
			log.setLevel(config.logLevel);
		};
		options = await preResolveOptions(inlineOptions,config,configEnv);
		const extraViteConfig = buildExtraViteConfig(options,config);;
		log.debug("additional vite config",extraViteConfig);;
		testΦ = configEnv.mode === "test";
		buildΦ = configEnv.mode === "production";
		devΦ = configEnv.mode === "development";
		return extraViteConfig;;
	};
	function configResolved(config){
		
		options = resolveOptions(options,config);;
		// patchResolvedViteConfig(config, options);
		requestParser = buildIdParser(options);;
		compileImba = createCompileImba(options);;
		viteConfig = config;;
		// // TODO deep clone to avoid mutability from outside?
		api.options = options;;
		return log.debug("resolved options",options);;
	};
	async function transform(code,id,opts){
		
		const ssr = !(!opts?.ssr);;
		const imbaRequest = requestParser(id,ssr);;
		
		if (!imbaRequest || imbaRequest.query.imba) { return };
		
		let compiledData;
		try {
			compiledData = await compileImba(imbaRequest,code,options);
		} catch (e) {
			
			cache.setError(imbaRequest,e);;
			throw toRollupError(e,options);;
		};
		logCompilerWarnings(
			imbaRequest,
			compiledData.compiled.warnings,
			options
		);
		cache.update(compiledData);
		if (compiledData.dependencies.length && options.server) {
			
			compiledData.dependencies.forEach(function(d) {
				
				return ensureWatchedFile(options.server.watcher,d,options.root);
			});
		};
		
		log.debug(("transform returns compiled js for " + (imbaRequest.filename)));
		if (imbaRequest.query.iife) {
			
			compiledData.compiled.js = await transformWithEsbuild(compiledData.compiled.js.code,imbaRequest.normalizedFilename.replace(".imba",".js"),{format: "iife"});
			compiledData.compiled.js.code = ("const body = " + JSON.stringify(compiledData.compiled.js.code) + "; export default \{body: body\}; export \{body\}");
		};
		
		return {
			...compiledData.compiled.js,
			meta: {
				vite: {
					lang: compiledData.lang
				}
			}
		};
	};
	
	function resolveId(id,importer,opts){
		
		let ssr = !(!opts?.ssr) || options.ssr;
		if (testΦ) { ssr = false };
		const imbaRequest = requestParser(id,ssr);
		if (id == allCssModuleId || id == "*?css" || id == "*") { return resolvedAllCssModuleId };
		if (imbaRequest?.query.imba) {
			
			if (imbaRequest.query.type === "style") {
				
				log.debug(("resolveId resolved virtual css module " + (imbaRequest.cssId)));
				return imbaRequest.cssId;
			};
			log.debug(("resolveId resolved " + id));
			return id;
		};
		if (ssr && id === "imba") {
			
			if (!resolvedImbaSSR) {
				
				resolvedImbaSSR = this.resolve("imba/server",undefined,{skipSelf: true}).then(function(imbaSSR) {
					
					log.debug("resolved imba to imba/server");
					return imbaSSR;
				});
			};
			return resolvedImbaSSR;
		};
		try {
			
			const resolved = resolveViaPackageJsonImba(id,importer,cache);
			if (resolved) {
				
				log.debug(("resolveId resolved $" + resolved + " via package.json imba field of " + id));
				return resolved;
			};
		} catch (e) {
			
			return log.debug.once(("error trying to resolve " + id + " from " + importer + " via package.json imba field "),e);
		};
	};
	
	function load(id,opts){
		
		const ssr = !(!opts?.ssr);
		const imbaRequest = requestParser(id,!!ssr);
		if (resolvedAllCssModuleId == id) {
			
			// if dev?
			return 'export default ".dev-ssr/all.css"';
		};
		if (imbaRequest) {
			
			const {filename: filename,query: query} = imbaRequest;
			if (query.imba && query.type === "style") {
				
				const _css = cache.getCSS(imbaRequest);
				if (_css) {
					
					log.debug(("load returns css for " + filename));
					return _css;
				} else {
					
					console.log(("cache empty: loading " + id));
				};
			};
			
			
			if (viteConfig.assetsInclude(filename)) {
				
				log.debug(("load returns raw content for " + filename));
				return fs.readFileSync(filename,"utf-8");
			};
		};
	};
	function configureServer(server){
		
		options.server = server;
		// Breaks tests, do not use until Vite has a proper fix
		// server.middlewares.use do(req, res, next)
		// 	const pathname = url.parse(req.url).pathname;
		// 	if pathname.endsWith ".imba"
		// 		req.url += "?import"
		// 		res.setHeader "Content-Type", "application/javascript"
		// 	next()
		// server.middlewares.use do(req, res, next)
		// 	const pathname = url.parse(req.url).pathname
		// 	if pathname.endsWith ".imba"
		// 		req.url = injectQuery req.url, 'import'
		// 		res.setHeader('Content-Type', 'application/javascript')
		// 	next!
		return setupWatchers(options,cache,requestParser);
	};
	/**
	@param {HmrContext} ctx
	*/
	async function handleHotUpdate(ctx){
		
		if (!options.hot || !options.emitCss) {
			
			return;
		};
		const imbaRequest = requestParser(ctx.file,false,ctx.timestamp);
		if (imbaRequest) {
			
			try {
				
				return await handleImbaHotUpdate(compileImba,ctx,imbaRequest,cache,options);
			} catch (e) {
				
				throw toRollupError(e,options);
			};
		};
	};
	
	// def handleHotUpdate(context\HmrContext)
	// 	console.log "context", context
	// 	const normalizedFilename = normalize(context.file, options.root);
	// 	const old-js-file = cache.getJS({ssr: no, normalizedFilename})
	// 	const new-js-file = await context.read!
	
	// 	debugger	
	// 	if new-js-file == old-js-file.code
	// 		return context.modules.filter do $1.url.includes "type=style"
	
	plugins.push(svgPlugin());
	plugins.push(
		{name: "vite-plugin-imba",
		enforce: "pre",
		api: api,
		config: config,
		configResolved: configResolved,
		transform: transform,
		load: load,
		resolveId: resolveId,
		configureServer: configureServer,
		handleHotUpdate: handleHotUpdate}
	);
	// plugins.push iifePlugin!
	
	return plugins;
	
};
