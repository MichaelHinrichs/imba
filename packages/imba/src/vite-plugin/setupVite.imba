import np from 'path'
import * as Vite from 'vite'
import url from 'node:url'
import nfs from 'node:fs'
import getport from 'get-port'
import { __served__ } from '../imba/utils'
import {getConfigFilePath} from '../utils/vite'
let vite

export def setupVite(srv, options, cb)
	global.__vite__ = yes
	if typeof options == 'string'
		options = {mode: options}
	elif typeof options == 'function'
		options = {mode: "development"}
	const prod? = options.mode == "production" or options.prod == yes
	unless vite
		const port = await getport(port: getport.makeRange(24000, 26000))

		const serverConfigFile = getConfigFilePath("server")
		let {default: serverConfig} = await import(serverConfigFile)
		if typeof serverConfig == "function"
			serverConfig = serverConfig({command: "build", mode: "production"})

		let serverOptions = options.serverOptions or { server: { hmr: {port} } }

		let clientConfigFile = getConfigFilePath "client"
		const vite-options = {
			appType: "custom"
			configFile: clientConfigFile
			...serverOptions,
			server: {
				...serverConfig.server,
				...serverOptions.server,
				middlewareMode: true
			}
		}

		vite = await Vite.createServer vite-options

	const dist = np.join(vite.config.root, vite.config.build.outDir)
	if prod?
		vite.close().then do vite = null
		global.__vite_manifest__ ||= JSON.parse nfs.readFileSync("{dist}/manifest.json", 'utf-8')
		cb(dist) if cb
		return dist
	else
		srv.use do(req, res, next)
			__served__.add req.url
			next()
		srv.use vite.middlewares
		srv.use do(req, res, next)
			__served__.delete req.url
			next()
	return dist
