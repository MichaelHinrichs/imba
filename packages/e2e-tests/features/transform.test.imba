import {generateImbaCode} from '../../imba/src/vite-plugin/cucumber-plugin'
import np from 'node:path'

def transform(path)
	path = np.resolve np.join('features', path)
	await generateImbaCode(path)

# It makes sense to write these tests manually
# if this is broken then feature tests won't run
describe "Cucumber plugin " do
	test "Add.feature" do
		const code = await transform("add.feature")
		expect(code).toMatchSnapshot!
	test "scenario-outline feature" do
		const code = await transform("scenario-outline.feature")
		expect(code).toMatchSnapshot!
