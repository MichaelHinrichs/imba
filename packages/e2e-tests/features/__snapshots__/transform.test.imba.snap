// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cucumber plugin  > Add.feature 1`] = `
"const modules = import.meta.glob('./step_definitions/**/*.imba', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
const decorators = {
}


beforeAll do(suite)
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	elName = 'Background: Sets initial value for the sum'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c0 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario: Adding two numbers works with background'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c1 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario: Supports multiple scenarios'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c2 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}

	s = global.Steps.find('the initial value is 10', 'Given').stepDefinition
	# find the context for the current test this beforeEach hook is running before
	c = Object.keys(contexts).find(do 
		contexts[$1].name == suite.meta.name or contexts[$1].name == suite.meta.suite..suite..name 
	)
	# apply the step
	await s.target[s.fname].apply(
		contexts[c].context,
		[s.cucumberExpression.match('the initial value is 10').map(do $1.getValue!)]
	)


describe 'Feature: Addition' do
	test('Scenario: Adding two numbers works with background')	do

		s = global.Steps.find('I add 12 and 4', 'Given').stepDefinition
		await s.target[s.fname]
			.apply(
				contexts.c1.context,
				# variables
				[s.cucumberExpression.match('I add 12 and 4').map(do $1.getValue!)]
			)

		s = global.Steps.find('I should obtain 26', 'Then').stepDefinition
		await s.target[s.fname]
			.apply(
				contexts.c1.context,
				# variables
				[s.cucumberExpression.match('I should obtain 26').map(do $1.getValue!)]
			)
	test('Scenario: Supports multiple scenarios')	do

		s = global.Steps.find('I add 12 and 10', 'Given').stepDefinition
		await s.target[s.fname]
			.apply(
				contexts.c2.context,
				# variables
				[s.cucumberExpression.match('I add 12 and 10').map(do $1.getValue!)]
			)

		s = global.Steps.find('I should obtain 32', 'Then').stepDefinition
		await s.target[s.fname]
			.apply(
				contexts.c2.context,
				# variables
				[s.cucumberExpression.match('I should obtain 32').map(do $1.getValue!)]
			)
"
`;

exports[`Cucumber plugin  > scenario-outline feature 1`] = `
"const modules = import.meta.glob('./step_definitions/**/*.imba', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
const decorators = {
}


beforeAll do(suite)
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	elName = 'Background: '
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c0 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario Outline: Adding two numbers works on many cases'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c1 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}

	s = global.Steps.find('the initial value is 0', 'Given').stepDefinition
	# find the context for the current test this beforeEach hook is running before
	c = Object.keys(contexts).find(do 
		contexts[$1].name == suite.meta.name or contexts[$1].name == suite.meta.suite..suite..name 
	)
	# apply the step
	await s.target[s.fname].apply(
		contexts[c].context,
		[s.cucumberExpression.match('the initial value is 0').map(do $1.getValue!)]
	)


describe 'Feature: Scenario outline' do
	describe 'Scenario Outline: Adding two numbers works on many cases' do
		describe 'Examples group 0: Basic examples' do
			let text
			test 'Example 0: 1, 2, 3' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 1)
				text = text.replace('<b>', 2)
				text = text.replace('<result>', 3)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 1)
				text = text.replace('<b>', 2)
				text = text.replace('<result>', 3)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 1: 12, 2, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 12)
				text = text.replace('<b>', 2)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 12)
				text = text.replace('<b>', 2)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 2: 13, 1, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 13)
				text = text.replace('<b>', 1)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 13)
				text = text.replace('<b>', 1)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 3: 14, 0, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 14)
				text = text.replace('<b>', 0)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 14)
				text = text.replace('<b>', 0)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 4: 15, -1, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 15)
				text = text.replace('<b>', -1)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 15)
				text = text.replace('<b>', -1)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 5: 16, -2, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 16)
				text = text.replace('<b>', -2)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 16)
				text = text.replace('<b>', -2)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 6: 17, -3, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 17)
				text = text.replace('<b>', -3)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 17)
				text = text.replace('<b>', -3)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
			test 'Example 7: 18, -4, 14' do
				text = 'I add <a> and <b>'
				text = text.replace('<a>', 18)
				text = text.replace('<b>', -4)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Given').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
				text = 'I should obtain <result>'
				text = text.replace('<a>', 18)
				text = text.replace('<b>', -4)
				text = text.replace('<result>', 14)
				s = global.Steps.find(text, 'Then').stepDefinition
				await s.target[s.fname]
					.apply(
						contexts.c1.context,
						# variables
						[s.cucumberExpression.match(text).map(do $1.getValue!)]
					)
				   
"
`;
